Add pandas io

choldgraf opened the pull request:
Adds read / write functionality for pandas dataframes (maybe series as well)

choldgraf commented:
OK, here's a first pass at the `_triage_write` function. The following works on my computer:

```python
# Create some fake data and write it to an HDF5 file
data = [np.random.randn(100, 10) for _ in range(2)]
df = pd.DataFrame(np.random.randn(100, 2), columns=['a', 'b'])

# Define some structures of varying types
data_list = [data, df]
data_tuple = (data, df)
data_dict = {'reg': data, 'mydf': df}
data_nested_dict = {'reg': data, 'mydf': {'another_df': df, 'yet_another': df}}
data_nested_list = [data, [df, df]]
alldata = [data_list, data_tuple, data_dict,
           data_nested_dict, data_nested_list]

# Write the data to file
out_hdf = './data/test_hdf_h5io.h'
h5io.write_hdf5(out_hdf, alldata, overwrite=True)
```
Now I'll inspect the file to show how it looks inside:
```python
# Now open the HDF5 file to see how it's stored
fpath = './data/test_hdf_h5io.h'
with h5py.File(fpath, 'r') as fid:
    print(fid.keys())
    print('---\n')

    rt = fid['h5io']
    print(rt.keys())
    print('---\n')

    for key in rt.keys():
        print(''.join([key, ': ', str(rt[key].keys())]))
    print('---\n')

    print(rt['idx_3']['key_mydf']['key_another_df'].keys())
```
This outputs...
```bash
[u'h5io']
---

[u'idx_0', u'idx_1', u'idx_2', u'idx_3', u'idx_4']
---

idx_0: [u'idx_0', u'idx_1']
idx_1: [u'idx_0', u'idx_1']
idx_2: [u'key_mydf', u'key_reg']
idx_3: [u'key_mydf', u'key_reg']
idx_4: [u'idx_0', u'idx_1']
---

[u'axis0', u'axis1', u'block0_items', u'block0_values']
```
Reading works in pandas just fine.
```python
# We can easily read in the data in pandas, but need to add it in h5io
df = pd.read_hdf(fpath, 'h5io/idx_3/key_mydf/key_another_df')
print(df)
```
However, I still need to figure out how to make this work with _triage_read. It could be a bit tricky because we're returning objects that aren't completely populated yet. I wonder if we could store a pointer to a place in the output object where we want the dataframe to go, and then recursively find those pointers and populate them with their actual data. I can work on this if you think this is headed in the right direction.

(and I'm sure I probably broke a flake or two but I'll fix 'em before a final commit)

choldgraf commented:
Ah - and another challenge here is that it currently tries to import pandas at the end of _triage_write. But I  guess this could be made a non-requirement pretty easily.

Eric89GXL commented:
> I still need to figure out how to make this work with _triage_read. It could be a bit tricky because we're returning objects that aren't completely populated yet. I wonder if we could store a pointer to a place in the output object where we want the dataframe to go, and then recursively find those pointers and populate them with their actual data. I can work on this if you think this is headed in the right direction.

I you can add a `None` as a placeholder if you want. But I wouldn't then go "search and find" them at the end -- `_triage_read` itself is a "search and find" operator, so it should maintain (append to) a list of the locations that must be populated, just like the writing function does. It  makes sense in my head at least. Make sense to you?

> Ah - and another challenge here is that it currently tries to import pandas at the end of _triage_write. But I guess this could be made a non-requirement pretty easily.

Yeah you can add a `try/except` somewhere.

choldgraf commented:
Yep - that's the kind of thing I was imagining, I don't think I explained myself well. But I imagine something similar to the _triage_write function where at the end it says "give me a list of data, and a position to write that data to". After triage_read we'd have a function that says "give me a list of data positions, and a place to put the data after I read it out".

Eric89GXL commented:
Yeah exactly

choldgraf commented:
It was actually pretty easy to add the reading functionality since multiple processes can open a file in read-only mode. There should be both read/write functionality now, as well as both DataFrame and Series functionality. I added a try/except bit for pandas checks. WDYT?

choldgraf commented:
ps: not sure about the travis build...it's giving anaconda update erros

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -123,19 +145,34 @@ def _triage_write(key, value, root, comp_kw, where):
+                    title = 'pd_series'
+                rootname = root.name
+                cleanup_data.append({title: (rootname, key, value)})
+        except ImportError:
```
I think I've had `pandas` imports crap out for other reasons, so let's do `except Exception`. It's also better to have the expected-offending line be the only one in the `try` clause so this might be a better approach.

1. Put this at the top of the file:
    ```
try:
    from pandas import DataFrame, Series
except Exception:
    DataFrame = Series = type(None)  # safe b/c we checked already
    ```

2. Then in both triage functions, you can just do:

    ```
# safe if these are type(None) because we looked for those already above,
# so on systems without Pandas this will never evaluate to True
elif isinstance(x, (DataFrame, Series))
    ...
else:
    raise TypeError(...)
```
Make sense?

Eric89GXL replied:
It can go next to the equivalent thing for `scipy`:

https://github.com/h5io/h5io/blob/master/h5io/_h5io.py#L12

Eric89GXL commented:
Can you add `-p /home/travis/miniconda` to the `miniconda.sh` call in `.travis.yml`?

Eric89GXL commented:
Never mind, I'll do that. Needs some other updates, too.

Eric89GXL commented:
I'll update Travis to have Pandas while I'm in there

Eric89GXL commented:
Okay you should be good to go

choldgraf commented:
Ah that's a clever idea. OK change made, I see that's what you did with `sparse` as well.

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -86,23 +99,37 @@ def write_hdf5(fname, data, overwrite=False, compression=4,
-
-
-def _triage_write(key, value, root, comp_kw, where):
+        cleanup_data = []
```
you can just scrap this line...

Eric89GXL replied:
Oh wait. Actually, leave this line in.

choldgraf replied:
What if I took that line out and passed nothing on the first call, and then did 

```python
def _triage_write(key, value, root, comp_kw, where, cleanup_data=[]):
```

choldgraf replied:
or kept the bit about `= [] if None else cleanup_data`

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -123,19 +150,32 @@ def _triage_write(key, value, root, comp_kw, where):
+        else:
+            err_str = 'unsupported type %s (in %s)' % (type(value), where)
+            raise TypeError(err_str)
+    return cleanup_data
```
you don't need to return anything as it's modified in place

Eric89GXL replied:
Note that there's nothing functionally wrong with what you did, but it's cleaner not to return anything because it makes it clearer it's modified in place. In other words, the return / reassignment doesn't really do anything. By omitting it, then, you make it clearer the object must be getting modified in place.

Eric89GXL replied:
still don't need this return

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -86,23 +99,37 @@ def write_hdf5(fname, data, overwrite=False, compression=4,
-
-def _triage_write(key, value, root, comp_kw, where):
+        cleanup_data = []
+        cleanup_data = _triage_write(title, data, fid, comp_kw,
```
and get rid of the `cleanup_data = ` here (it's modified in place)

choldgraf replied:
right good point

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -86,23 +99,37 @@ def write_hdf5(fname, data, overwrite=False, compression=4,
+
+
+def _triage_write(key, value, root, comp_kw, where, cleanup_data=None):
+    cleanup_data = [] if cleanup_data is None else cleanup_data
```
you shouldn't need this line -- `cleanup_data` should never be `None`, right?

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -86,23 +99,37 @@ def write_hdf5(fname, data, overwrite=False, compression=4,
                 raise TypeError('All dict keys must be strings')
-            _triage_write('key_{0}'.format(key), sub_value, sub_root, comp_kw,
-                          where + '["%s"]' % key)
+            cleanup_data = _triage_write(
```
again, don't need `cleanup_data = ` but just pass it to the function

Eric89GXL commented on h5io/_h5io.py:
```diff
@@ -86,23 +99,37 @@ def write_hdf5(fname, data, overwrite=False, compression=4,
         for vi, sub_value in enumerate(value):
-            _triage_write('idx_{0}'.format(vi), sub_value, sub_root, comp_kw,
-                          where + '[%s]' % vi)
+            cleanup_data = _triage_write(
```
etc.

Eric89GXL commented:
It's getting there. Please update a test in the same way we do with `sparse`

choldgraf commented:
What do you want me to do about object_diff? I haven't looked into the function much so I'm not sure what it does exactly, but should I add some kind of support for pandas objects?

choldgraf commented:
nvm, I see that it's required in the checks - I'll try to cook up a comparison

choldgraf commented:
ok,  tests made - I pretty much did all the same stuff as was already performed on 'sparse'. I decided to leave in the cleanup_data = [] line because I think it makes it more clear what we're doing (rather than implicitly making this an empty list by not passing `cleanup_data` as a kwarg.

One problem - travis is breaking because pandas requires tables, so when the tests run on pandas it's throwing an error.

choldgraf commented:
oops, missed that one

Eric89GXL commented:
Well that's silly that `conda` doesn't resolve the `tables` dependency. You'll need to add `pytables` to the non-minimal deps in `.travis.yml` and to the install line in `appveyor.yml`

choldgraf commented:
hrm, now it says that it's non-mergeable

choldgraf commented:
ah I probably need to rebase because of the changes you made

Eric89GXL commented:
yep

choldgraf commented:
sigh, looks like it's returning python 3 errors now

Eric89GXL commented:
You have a `minimal` error, too. I think if you set them both to `type(None)` it might fix that one

Eric89GXL commented:
On py3k `.keys()` is a generator, so `list(var.keys())[0]` should be the equivalent. But keep in mind that the order of dicts is not guaranteed (if that is a dict) -- do you know there will be exactly one key, or is it not a dict?

choldgraf commented:
For the minimal error I'm a bit confused what you mean. Do you mean to set DataFrame and Series to type(None) and then do a check for `if DataFrame is not type(None)`? Why wouldn't this also fail for sparse?

For the dictionary error, it shouldn't be a big deal because there should only be a single value in there anyway.

Eric89GXL commented:
Looking at the code for `Sparse`:
```
elif sparse is not None and isinstance(value, sparse.csc_matrix):
```
So you could alternatively follow that pattern if you want I guess.

choldgraf commented:
hmm - well I changed them to type(None) and added those to the checks in the test code as well. fingers crossed. (ps I'm heading to lunch really quickly, will fix any errors in like 30-60.

choldgraf commented:
Sorry I had to go so abruptly - had a lunch planned that I was already late for. Made a last couple of changes and I think it's building correctly now. Though coveralls isn't happy...

choldgraf commented:
ok cools like they both pass - not sure what to do about coveralls though...

Eric89GXL commented:
Thanks @choldgraf!

Eric89GXL commented:
I'm not worried about the -0.3% for coveralls :)
